agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java

```diff
@@ -0,0 +1,283 @@
+          span.setAttribute(MoreTags.HTTP_FRAGMENT, url.getFragment());
+        }
+      }
+    } catch (final Exception e) {
+      log.debug("Error tagging url", e);
+    }
+    // TODO set resource name from URL.
+  }
+  /**
+   * This method is used to generate an acceptable span (operation) name based on a given method
+   * reference. Anonymous classes are named based on their parent.
+   */
+  protected String spanNameForMethod(final Method method) {
+    return spanNameForClass(method.getDeclaringClass()) + "." + method.getName();
+  }
+
+  /**
+   * This method is used to generate an acceptable span (operation) name based on a given class
+   * reference. Anonymous classes are named based on their parent.
+   */
+  public String spanNameForClass(final Class clazz) {
+    if (!clazz.isAnonymousClass()) {
+      return clazz.getSimpleName();
+    }
+    String className = clazz.getName();
+    if (clazz.getPackage() != null) {
+      final String pkgName = clazz.getPackage().getName();
+      if (!pkgName.isEmpty()) {
+        className = clazz.getName().replace(pkgName, "").substring(1);
+      }
+    }
+    return className;
+  }
+
+  protected void onError(final Span span, final Throwable throwable) {
+    addThrowable(span, unwrapThrowable(throwable));
+  }
+
+  // TODO semantic attributes
+  public static void addThrowable(final Span span, final Throwable throwable) {
+    span.setAttribute(MoreTags.ERROR_MSG, throwable.getMessage());
+    span.setAttribute(MoreTags.ERROR_TYPE, throwable.getClass().getName());
+
+    final StringWriter errorString = new StringWriter();
+    throwable.printStackTrace(new PrintWriter(errorString));
+    span.setAttribute(MoreTags.ERROR_STACK, errorString.toString());
+  }
+
+  public Span getCurrentSpan() {
+    return tracer.getCurrentSpan();
+  }
+
+  // TODO should end methods remove SPAN attribute from request as well?
+
+  public void end(SpanWithScope spanWithScope, int responseCode) {
+    if (spanWithScope == null) {
+      return;
+    }
+
+    final Span span = spanWithScope.getSpan();
+    if (span != null) {
+      end(span, responseCode);
+    }
+
+    spanWithScope.closeScope();
+  }
+
+  public void end(Span span, int responseStatus) {
+    setStatus(span, responseStatus);
+    span.end();
+  }
+
+  public void endExceptionally(
+      SpanWithScope spanWithScope, Throwable throwable, int responseStatus) {
+    if (spanWithScope == null) {
+      return;
+    }
+
+    final Span span = spanWithScope.getSpan();
+    if (span != null) {
+      endExceptionally(span, throwable, responseStatus);
+    }
+
+    spanWithScope.closeScope();
+  }
+
+  public void endExceptionally(Span span, Throwable throwable, int responseStatus) {
+    if (responseStatus == 200) {
+      // TODO I think this is wrong.
+      // We must report that response status that was actually sent to end user
+      // We may change span status, but not http_status attribute
```

Comment:
I think this code is correct, short of checking `response.isCommitted()`.

From the servlet spec:

> If the exception isn't caught and handled, and the response hasn’t been committed, the status code MUST be set to 500

Reply:
But we don't have `response` here any more :) By your own suggestion, we now only have response status code :)

Reply:
Good point, since this behavior (setting status to `500` if response is not committed) is specific to the servlet instrumentation, maybe we should pull this out into the caller?

Reply:
See https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/469/files#r434780039